### PR TITLE
applications: machine_learning: Add machine learning state

### PR DESCRIPTION
**Pull request by MarekPieta**
Change introduces machine learning state. User can switch between forwarding data and running the machine learning model using a predefined button.

### CODE REVIEW

**Review by pdunaj — APPROVED** [review]

File: applications/machine_learning/src/modules/ml_runner.c [hunk 1]
```diff
@@ -48,20 +54,97 @@ static void result_ready_cb(int err)
...
+static void report_error(void)
+{
+	state = STATE_ERROR;
+	module_set_state(MODULE_STATE_ERROR);
+}
 
-		err = ei_wrapper_start_prediction(SHIFT_WINDOWS, SHIFT_FRAMES);
+static int buf_cleanup(void)
+{
+	bool cancelled;
```
Comment (pdunaj) [inline applications/machine_learning/src/modules/ml_runner.c:72]: set it to zero to appease coverity

File: applications/machine_learning/src/modules/ml_runner.c [hunk 2]
```diff
@@ -48,20 +54,97 @@ static void result_ready_cb(int err)
...
+	int err;
+	size_t window_shift;
+	size_t frame_shift;
+
+	if (ml_control & ML_RUNNING) {
+		return;
+	}
+
+	if (ml_control & ML_CLEANUP_REQUIRED) {
+		if (buf_cleanup()) {
```
Comment (pdunaj) [inline applications/machine_learning/src/modules/ml_runner.c:109]: I would prefer
```
err = buf_cleanup();
if (err) {
   return;
}
```

File: applications/machine_learning/src/modules/ml_runner.c [hunk 3]
```diff
@@ -48,20 +54,97 @@ static void result_ready_cb(int err)
...
+
+	if (ml_control & ML_FIRST_PREDICTION) {
+		window_shift = 0;
+		frame_shift = 0;
+	} else {
+		window_shift = SHIFT_WINDOWS;
+		frame_shift = SHIFT_FRAMES;
+	}
+
+	err = ei_wrapper_start_prediction(window_shift, frame_shift);
```
Comment (pdunaj) [inline applications/machine_learning/src/modules/ml_runner.c:122]: `int err`

File: applications/machine_learning/src/modules/ml_runner.c [hunk 4]
```diff
@@ -19,26 +20,31 @@ LOG_MODULE_REGISTER(MODULE, CONFIG_ML_APP_ML_RUNNER_LOG_LEVEL);
 #define SHIFT_WINDOWS		CONFIG_ML_APP_ML_RUNNER_WINDOW_SHIFT
 #define SHIFT_FRAMES		CONFIG_ML_APP_ML_RUNNER_FRAME_SHIFT
 
+#define ML_STATE_CONTROL	IS_ENABLED(CONFIG_ML_APP_ML_STATE_EVENTS)
```
Comment (pdunaj) [inline applications/machine_learning/src/modules/ml_runner.c:23]: This name is very confusing as it blends with names in this new enum. See line 198 to know what I mean.

File: applications/machine_learning/src/modules/ml_runner.c [hunk 5]
```diff
@@ -102,21 +213,16 @@ static bool event_handler(const struct event_header *eh)
...
+	}
 
-			if (!init()) {
-				module_set_state(MODULE_STATE_READY);
-				state = STATE_ACTIVE;
-			} else {
-				report_error();
-			}
-		}
+	if (ML_STATE_CONTROL && is_ml_state_event(eh)) {
```
Comment (pdunaj) [inline applications/machine_learning/src/modules/ml_runner.c:220]: can we write it as
```
if (CONDITION &&
    is_ml_state_event(eh)) {
...
```

I find it easier to follow as you don't have to look right to see what event is handled in this block.

**Review by pdunaj — CHANGES_REQUESTED** [review]

File: applications/machine_learning/src/modules/ml_runner.c (hunk 1 above)
Comment (MarekPieta) [reply applications/machine_learning/src/modules/ml_runner.c:72]: done

File: applications/machine_learning/src/modules/ml_runner.c (hunk 2 above)
Comment (MarekPieta) [reply applications/machine_learning/src/modules/ml_runner.c:109]: done

File: applications/machine_learning/src/modules/ml_runner.c (hunk 3 above)
Comment (MarekPieta) [reply applications/machine_learning/src/modules/ml_runner.c:122]: Used to store results of buf_cleanup a few lines above

File: applications/machine_learning/src/modules/ml_runner.c (hunk 4 above)
Comment (MarekPieta) [reply applications/machine_learning/src/modules/ml_runner.c:23]: You are right. Changed.

File: applications/machine_learning/src/modules/ml_runner.c (hunk 5 above)
Comment (MarekPieta) [reply applications/machine_learning/src/modules/ml_runner.c:220]: Changed.

**Review by pdunaj — APPROVED** [review]

**Review by emob-nordic — APPROVED** [review]

**Review by nvlsianpu — APPROVED** [review]

**Pull request merged**